### PR TITLE
CSharp Service Emitter

### LINKED ISSUE #3141: [Service CodeGen]: Complete existing base type support in CSharp Service Emitter
- [x] Finish representation and serialization of unions
- [x] Finish support and tests for model visibility

### PR DESCRIPTION
**Pull request by markcowl**
Fixes #3141

### CODE REVIEW

**Comment by markcowl** [comment]
Fixes #3141

File: packages/http-server-csharp/src/lib.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      nullable: true,
+      default: "all",
+      description:
+        "Chooses which service artifacts to emit. choices include 'models' or 'all' artifacts.",
+    },
+  },
+  required: [],
+};
+
+export const libDef = {
```
Comment (timotheeguerin) [inline packages/http-server-csharp/src/lib.ts:32]:
```suggestion
export const $lib = createTypeSpecLibrary({
```
nit i would combine that with the next line so you can autocomplete for the structure

File: packages/http-server-csharp/test/diagnostic.test.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
+import { BasicTestRunner, expectDiagnostics } from "@typespec/compiler/testing";
+import { beforeEach, describe, it } from "vitest";
+import { createCSharpServiceEmitterTestRunner, getStandardService } from "./test-host.js";
+describe("http-server-csharp: diagnostics", () => {
```
Comment (timotheeguerin) [inline packages/http-server-csharp/test/diagnostic.test.ts:4]: don't need to have top level describe with vitest just adds an extra nesting in the tree explorer

File: packages/http-server-csharp/package.json [hunk 3]
```diff
@@ -0,0 +1,80 @@
+{
+  "name": "@typespec/http-server-csharp",
+  "version": "0.58.0",
```
[the remark this answers is not in the document]
Comment (markcowl) [reply packages/http-server-csharp/package.json:3]: I was thinking that, but it looks like it is using 0.58.0: https://github.com/microsoft/typespec/blob/main/packages/http-server-javascript/package.json#L3

If we want to add -alpha, I am not opposed

[the remark this answers is not in the document]
Comment (timotheeguerin) [reply packages/http-server-csharp/package.json:3]: that was a mistake with the last release where it bumped it, I deprecated that version and will was gonna make a pr to revert back.

I think its better specially if the goal is to get something out for now but not have the same try not to break policy

File: packages/http-server-csharp/package.json [hunk 4]
```diff
@@ -0,0 +1,77 @@
+{
+  "name": "@typespec/http-server-csharp",
+  "version": "0.58.0-alpha.1",
+  "author": "Microsoft Corporation",
+  "description": "TypeSpec service code generator for c-sharp",
+  "homepage": "https://typespec.io",
+  "readme": "https://github.com/microsoft/typespec/blob/main/README.md",
```
Comment (timotheeguerin) [inline packages/http-server-csharp/package.json:7]:
```suggestion
  "readme": "https://github.com/microsoft/typespec/blob/main/packages/http-server-csharp/README.md",
```

File: packages/http-server-csharp/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "../../tsconfig.base.json",
+  "references": [{ "path": "../compiler/tsconfig.json" }, { "path": "../rest/tsconfig.json" }],
```
Comment (timotheeguerin) [inline packages/http-server-csharp/tsconfig.json:3]: that list should match the workspace deps, missing versioning and http

**Comment by connorjs** [comment]
Happy end of July/start of July! 🎉

Just wanted to say how I’m excited to try this out and to say thank you for all the work on TypeSpec. I continue to watch the project grow 😃

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**